# Business Understanding

## Overview

# Data Understanding

The data being used is from a Credit Bereau. and it consists of:

**Application Data:**
* **Application_ID**: Unique Application Number
* **Application_Date**: Date of application
* **Loan_Approval_Status**: Approval status of loan
* **Age_at_Application**: Age in Years at point of application
* **Marital_Status**: Marital status at point of application
* **Gender**: Sex
* **Own_Rent_Home**: Residential status of home ownership
* **Occupation_Professional_Category**: Proffessional Category
* **Type_of_Business_Industry_of_Employment**: Industry of Employment
* **Employment_Status**: Employment status
* **Time_at_Current_Employment_(Months)**: Time at current employment in months
* **Time_Employed_with_Previous_Employer_(Months)**: Time at previous employment in months
* **Number_of_Dependents**: Number of dependents
* **Loan_Purpose**: Purpose of Loan Applied
* **Monthly_Income**: Monthly Income in KES
* **Government_Employee**: Flag if government employee

**Contracts Data:**
* **Contract_ID**:	Loan Account Number
* **Application_ID**	Application Number related to the loan opened
* **Key**:	Number of Open Contracts related to the loan application Number
* **Loan_Type**:	Type of loan facility
* **Contract_Start_Date**:	Date of start for Loan facility
* **Expected_Contract_End_Date**:	Expected date of loan maturity
* **Actual_Contract_End_Date**:	Actual date when the contrcat was closed
* **Loan_Term**:	Tenure of loan in months
* **Total_Loan_Amount**:	Loan amount approved

**Contracts Snapshop data:**
* **Contract Id**:	Loan Account Number*
* **Snapshot Date**:	The End of Month when the data was reported
* **Outstanding Amount**:	The outstanding balance as at the reporting date
* **Due Amount**:	The amount unpaid as at the reporting date
* **Number of Days Past Due**:	The number of days past due as at that reporting date
* **Contract Status**:	The status of the contract

**CRB Data:**
* **Application_ID**:	Application Number related to the loan opened
* **Application_Date**:	Date of application
* **CRB Score**:	Credit Bureau Risk Score
* **CRB Grade**:	Credit Bureau Risk Grade

**Current and Savings Account Data:**
* **Account Number**:	Application Number related to the loan opened
* **Application ID**:	Application Number related to the loan opened
* **Month (YYYY-MM)**:	The reporting month when the data was calculated
* **Open Date**:	The date when the account was openned
* **Closed Date**:	The date when the account was closed
* **Average Balance**:	The monthly average balance
* **Overdraft Limit**:	Overdraft Limit
* **Number of Days in Debit during Month**:	Number of Days in Debit during Month
* **Number of Days in Credit during Month**: Number of Days in Credit during Month
* **Maximum End of Day Balance**:	Maximum End of Day Balance
* **Minimum End of Day Balance**:	Minimum End of Day Balance
* **Sum of Monthly Debit Transactions**: Sum of Monthly Debit Transactions
* **Sum of Monthly Credit Transactions**:	Sum of Monthly Credit Transactions
* **Last Transaction Date**:	Last Transaction Date
* **Sum of ATM Deposits**:	Sum of ATM Deposits
* **Number of ATM Deposits**:	Number of ATM Deposits
* **Sum of ATM Withdrawals**:	Sum of ATM Withdrawals
* **Number of ATM Withdrawals**:	Number of ATM Withdrawals
* **Sum of Cheque Debit Transactions**:	Sum of Cheque Debit Transactions
* **Number of Cheque Debit Transactions**:	Number of Cheque Debit Transactions
* **Sum of Bounced Cheque transactions**:	Sum of Bounced Cheque transactions
* **Number of Bounced Cheques**:	Number of Bounced Cheques
* **Sum of Debit Card POS Transactions**:	Sum of Debit Card POS Transactions
* **Number of Debit Card POS Transactions**:	Number of Debit Card POS Transactions
* **Sum of Over the Counter Withdrawals**:	Sum of Over the Counter Withdrawals
* **Number of Over the Counter withdrawals**:	Number of Over the Counter withdrawals
* **Account Type**:	Type of account
* **Account Status**:	Latest status of the account
* **Number of Joint Holders**:	Number of Persons owning the account
* **SnapshotMonth**:	The reporting month when the data was calculated

![image.png](..\creditdata\images\erpdiagram.png)


In [768]:
# Loading the relevant libraries

# Data pre-processsing libraries
import pandas as pd
import numpy as np
import re

# Metrics libraries

# Data modeling libraries

# 

### Application Data

In [769]:
# Load data
app_data = pd.read_csv("../creditdata/1. ApplicationData.csv")

# Check 2 top rows
app_data.head(2)

Application_ID Application_Date Loan_Approval_Status  Age_at_Application  \
0     C302100001       2015-09-15             Approved                39.0   
1     C302100002       2015-08-31             Approved                29.0   

  Marital_Status  Gender Own_Rent_Home Occupation_Professional_Category  \
0      Unmarried  Female           Own                              NaN   
1      Unmarried  Female        Family                              NaN   

  Type_of_Business_Industry_of_Employment Employment_Status  \
0                                   Other         Full-Time   
1                                     NaN         Full-Time   

   Time_at_Current_Employment_(Months)  \
0                                  6.0   
1                                 25.0   

   Time_Employed_with_Previous_Employer_(Months)  Number_of_Dependents  \
0                                            NaN                   2.0   
1                                            NaN                   0.0   

         Loan_Purpose  Monthly_Income Government_Employee  
0    Home Improvement         1500.59                 NaN  
1  Debt Consolidation         3769.36                 NaN

In [770]:
# Get data summary
app_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9492 entries, 0 to 9491
Data columns (total 16 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Application_ID                                 9492 non-null   object 
 1   Application_Date                               9492 non-null   object 
 2   Loan_Approval_Status                           9354 non-null   object 
 3   Age_at_Application                             9471 non-null   float64
 4   Marital_Status                                 7712 non-null   object 
 5   Gender                                         9488 non-null   object 
 6   Own_Rent_Home                                  8188 non-null   object 
 7   Occupation_Professional_Category               9361 non-null   object 
 8   Type_of_Business_Industry_of_Employment        7210 non-null   object 
 9   Employment_Status                              9357 

In [771]:
# Check Shape
app_data.shape

(9492, 16)

In [772]:
# Check null values
app_data.isnull().sum()

Application_ID                                      0
Application_Date                                    0
Loan_Approval_Status                              138
Age_at_Application                                 21
Marital_Status                                   1780
Gender                                              4
Own_Rent_Home                                    1304
Occupation_Professional_Category                  131
Type_of_Business_Industry_of_Employment          2282
Employment_Status                                 135
Time_at_Current_Employment_(Months)               232
Time_Employed_with_Previous_Employer_(Months)    9194
Number_of_Dependents                             1814
Loan_Purpose                                       42
Monthly_Income                                      0
Government_Employee                               181
dtype: int64

*Observation*

* The ApplicationData has $9,492$ records under $16$ columns.
* All the columns have null values apart from the *'Application_ID'*, *'Application_Date'*, and *'Monthly_Income'* columns.
* The 'Application_Date' column is  an object dtype. thiss needs to be changed to 'date' datatype

In [773]:
# app_data.Loan_Approval_Status.value_counts()

### Contracts Data

In [774]:
# Load data
contract_data = pd.read_csv("../creditdata/2. ContractsData.csv")

# Check 2 top rows
contract_data.head(2)

Contract_ID Application_ID  Key                Loan_Type  \
0  CONT100001     C302103426    1  Unsecured Personal Loan   
1  CONT100002     C302103426    2  Unsecured Personal Loan   

  Contract_Start_Date Expected_Contract_End_Date Actual_Contract_End_Date  \
0          2018-04-12                 2023-04-12                      NaN   
1          2018-04-12                 2023-04-12                      NaN   

       Loan_Term  Total_Loan_Amount  
0  60 - Month(s)           24529.89  
1  60 - Month(s)           35085.79

In [775]:
# show data summary
contract_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9374 entries, 0 to 9373
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Contract_ID                 9374 non-null   object 
 1   Application_ID              7598 non-null   object 
 2   Key                         9374 non-null   int64  
 3   Loan_Type                   9374 non-null   object 
 4   Contract_Start_Date         9374 non-null   object 
 5   Expected_Contract_End_Date  9374 non-null   object 
 6   Actual_Contract_End_Date    2593 non-null   object 
 7   Loan_Term                   9374 non-null   object 
 8   Total_Loan_Amount           9374 non-null   float64
dtypes: float64(1), int64(1), object(7)
memory usage: 659.2+ KB


In [776]:
# Check shape
contract_data.shape

(9374, 9)

In [777]:
# Check null values
contract_data.isnull().sum()

Contract_ID                      0
Application_ID                1776
Key                              0
Loan_Type                        0
Contract_Start_Date              0
Expected_Contract_End_Date       0
Actual_Contract_End_Date      6781
Loan_Term                        0
Total_Loan_Amount                0
dtype: int64

*Observation*
* The contract data is of $9$ columns with $9,374$ records
* The dates are in the object dtype, will need to be changed to date format
* The contracts data has null values on the *'Application_ID'* and *'Actual_Contract_End_Date'* columns.


### Contracts Snapshot data

In [778]:
# Load the dataset
ContractsSnapshot_data = pd.read_csv("../creditdata/3. ContractsSnapshotData.csv")

# Check top 2 rows
ContractsSnapshot_data.head(2)

Contract_ID Snapshot Date  Outstanding Amount  Due Amount  \
0  CONT109375    2018-12-31                0.00         0.0   
1  CONT109376    2006-08-31              215.08         0.0   

   Number of Days Past Due Contract Status  
0                      NaN          Closed  
1                      NaN          Active

In [779]:
# Check Data size
ContractsSnapshot_data.shape

(274371, 6)

In [780]:
# Data summary
ContractsSnapshot_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274371 entries, 0 to 274370
Data columns (total 6 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Contract_ID              274371 non-null  object 
 1   Snapshot Date            274371 non-null  object 
 2   Outstanding Amount       274371 non-null  float64
 3   Due Amount               274371 non-null  float64
 4   Number of Days Past Due  79158 non-null   float64
 5   Contract Status          274371 non-null  object 
dtypes: float64(3), object(3)
memory usage: 12.6+ MB


In [781]:
# Check null values
ContractsSnapshot_data.isnull().sum()

Contract_ID                     0
Snapshot Date                   0
Outstanding Amount              0
Due Amount                      0
Number of Days Past Due    195213
Contract Status                 0
dtype: int64

In [782]:
# Check data
ContractsSnapshot_data[10:20]

Contract_ID Snapshot Date  Outstanding Amount  Due Amount  \
10  CONT109376    2007-05-31               64.15        0.00   
11  CONT109376    2007-06-30               64.15       22.37   
12  CONT109376    2007-07-31               64.15       49.61   
13  CONT109376    2007-08-31               17.58       18.12   
14  CONT109376    2007-09-30                0.00       18.12   
15  CONT109377    2006-08-31              601.92        0.00   
16  CONT109377    2006-09-30              601.92        0.00   
17  CONT109377    2006-10-31              601.92       38.17   
18  CONT109377    2006-11-30              536.04        0.00   
19  CONT109377    2006-12-31              469.38        0.00   

    Number of Days Past Due Contract Status  
10                      NaN          Active  
11                      3.0          Active  
12                     34.0          Active  
13                     35.0          Active  
14                     65.0          Closed  
15                      NaN          Active  
16                      NaN          Active  
17                     26.0          Active  
18                      NaN          Active  
19                      NaN          Active

*Observation*
* Tha contractsSnapshot data has $6$ columns with $274,371$ rows.
* the date column is of object dtype. Will be changed to date dtype
* It has null values on the *'Number of Days Past Due'* column. On further checking, the null values are present on records that do not have *'Due Amount'*.

### CRB Data

In [783]:
# Load data
crb_data = pd.read_csv("../creditdata/4. CRB Data.csv")

# Check top 2 rows
crb_data.head(2)

Application_ID Application_Date  CRB Score CRB Grade
0     C302102137       2017-09-30        593         D
1     C302106175       2017-09-30        625         C

In [784]:
# Data summary
crb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6750 entries, 0 to 6749
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Application_ID    6750 non-null   object
 1   Application_Date  6750 non-null   object
 2   CRB Score         6750 non-null   int64 
 3   CRB Grade         6750 non-null   object
dtypes: int64(1), object(3)
memory usage: 211.1+ KB


In [785]:
# Check null values
crb_data.isnull().sum()

Application_ID      0
Application_Date    0
CRB Score           0
CRB Grade           0
dtype: int64

In [786]:
# Check size
crb_data.shape

(6750, 4)

*Observation*
* CRB Data has $4$ columns with $6,750$ records.
* The CRB Data has no null values.
* The *'Application_Date'* column is of object dtype, to be changed to date dtype.

In [787]:
#crb_data["CRB Grade"].value_counts()

### Current_and_Savings Data

In [788]:
# load data
curr_savings_data = pd.read_csv("../creditdata/5. Current and Savings Account Data.csv")

# Check top 2 rows
curr_savings_data.head(2)

ID Account_Number Application ID Month (YYYY-MM)   Open Date Closed Date  \
0   1    ACC10000001     C302100001      2015-09-01  2015-09-06         NaN   
1   2    ACC10000002     C302100001      2015-09-01  2015-09-06  2018-10-04   

   Average Balance  Overdraft Limit  Number of Days in Debit during Month  \
0          1986.09              0.0                                   0.0   
1            23.45              NaN                                   0.0   

   Number of Days in Credit during Month  ...  \
0                                   25.0  ...   
1                                   25.0  ...   

   Sum of Bounced Cheque transactions  Number of Bounced Cheques  \
0                                 NaN                        0.0   
1                                 NaN                        0.0   

   Sum of Debit Card POS Transactions  Number of Debit Card POS Transactions  \
0                                 NaN                                    0.0   
1                                 NaN                                    0.0   

  Sum of Over the Counter Withdrawals  Number of Over the Counter withdrawals  \
0                                 NaN                                     0.0   
1                                 NaN                                     0.0   

   Account Type  Account Status  Number of Joint Holders  SpshotMonth  
0       Savings          Active                      0.0   2015-09-01  
1       Savings          Closed                      0.0   2015-09-01  

[2 rows x 31 columns]

In [789]:
# Dataset summary
curr_savings_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857677 entries, 0 to 857676
Data columns (total 31 columns):
 #   Column                                  Non-Null Count   Dtype  
---  ------                                  --------------   -----  
 0   ID                                      857677 non-null  int64  
 1   Account_Number                          857677 non-null  object 
 2   Application ID                          857672 non-null  object 
 3   Month (YYYY-MM)                         857647 non-null  object 
 4   Open Date                               857647 non-null  object 
 5   Closed Date                             161748 non-null  object 
 6   Average Balance                         857647 non-null  float64
 7   Overdraft Limit                         753484 non-null  float64
 8   Number of Days in Debit during Month    857647 non-null  float64
 9   Number of Days in Credit during Month   857647 non-null  float64
 10  Maximum End of Day Balance              8576

In [790]:
# Check size
curr_savings_data.shape

(857677, 31)

In [791]:
# Check null values
# ~
curr_savings_data.isnull().sum().sort_values()

ID                                             0
Account_Number                                 0
Application ID                                 5
Account Status                                30
Account Type                                  30
Number of Over the Counter withdrawals        30
Number of Debit Card POS Transactions         30
Number of Bounced Cheques                     30
Number of Cheque Debit Transactions           30
Number of ATM Withdrawals                     30
Number of ATM Deposits                        30
Number of Joint Holders                       30
SpshotMonth                                   30
Minimum End of Day Balance                    30
Maximum End of Day Balance                    30
Number of Days in Credit during Month         30
Month (YYYY-MM)                               30
Number of Days in Debit during Month          30
Average Balance                               30
Open Date                                     30
Sum of Monthly Credi

*Observation*
* The Current & Savings data has $31$ columns with $857,677$ records
* This data has null values on all the columns apart from the *'ID'* and *'Account_Number'* columns
* The date columns are also of object dtype. To be changed to date dtype

In [792]:
curr_savings_data['Application ID'].value_counts()

C302102519    860
C302100587    860
C302103114    760
C302102862    745
C302103983    725
             ... 
C302109122      1
C302109064      1
C302108690      1
C302108802      1
C302109322      1
Name: Application ID, Length: 9492, dtype: int64

* **ApplicationData.csv**: Loan_Approval_Status, Monthly_Income, Number_of_Dependents, Time_at_Current_Employment_(Months), Employment_Status
* **ContractsData.csv**: Loan_Type, Loan_Term, Total_Loan_Amount
* **ContractsSnapshotData.csv**: Outstanding Amount, Number of Days Past Due, Contract Status
* **CRB Data.csv**: CRB Score, CRB Grade
* **Current and Savings Account Data.csv**: Average Balance, Overdraft Limit, Sum of Monthly Credit Transactions, Number of Bounced Cheques


## Otheer

In [793]:
app_data.columns

Index(['Application_ID', 'Application_Date', 'Loan_Approval_Status',
       'Age_at_Application', 'Marital_Status', 'Gender', 'Own_Rent_Home',
       'Occupation_Professional_Category',
       'Type_of_Business_Industry_of_Employment', 'Employment_Status',
       'Time_at_Current_Employment_(Months)',
       'Time_Employed_with_Previous_Employer_(Months)', 'Number_of_Dependents',
       'Loan_Purpose', 'Monthly_Income', 'Government_Employee'],
      dtype='object')

In [794]:
rel_app_data = app_data[["Loan_Approval_Status", "Monthly_Income", "Number_of_Dependents", "Time_at_Current_Employment_(Months)", "Employment_Status"]]

### Basic Overview - data EDA

In [795]:
# Check dataset shape (df.shape)


In [796]:
# Check data types (df.dtypes)


In [797]:
#Identify unique values in categorical columns (df.nunique())


### Summary statistics

In [798]:
# Numerical Data: Check for mean, median, min, max, and standard deviation .describe()

In [799]:
# Numerical: Check skewness df.skew()

In [800]:
# Categorical data: Frequency distribution (df["colomn"].value_counts())

# Data Cleaning / Preparation

Understand the data
1. Drop irrelevant columns
2. Handle duplicates
3. Handle missing data
4. Convert data types (e.g., datetime, categorical)
5. Normalize/encode data
6. Merge datasets (if applicable)
7. Feature engineering

## Having the right Data type
Let us change the dates on all the datasets to be of type datetime. But before that, i noticed some dates are not correctly recorded e.g. '*202-1-03-06*' which i think should be '*2021-03-06*'.
So we will format it too.

In [801]:
app_data[9446:9456]

Application_ID Application_Date   Loan_Approval_Status  \
9446     C302109447      202-1-03-06                    NaN   
9447     C302109448       2018-07-05                Decline   
9448     C302109449      202-1-02-21  Customer Cancellation   
9449     C302109450       2018-08-22  Customer Cancellation   
9450     C302109451       2018-09-19  Customer Cancellation   
9451     C302109452      202-1-02-27                    NaN   
9452     C302109453       2018-06-17  Customer Cancellation   
9453     C302109454       2018-11-25                Decline   
9454     C302109455       2017-04-22  Customer Cancellation   
9455     C302109456      202-1-08-12                    NaN   

      Age_at_Application Marital_Status  Gender Own_Rent_Home  \
9446                22.0            NaN  Female           NaN   
9447                31.0            NaN    Male           NaN   
9448                24.0            NaN    Male           NaN   
9449                37.0            NaN  Female           NaN   
9450                27.0            NaN    Male           NaN   
9451                39.0            NaN  Female           NaN   
9452                29.0            NaN    Male           NaN   
9453                37.0            NaN    Male           NaN   
9454                22.0            NaN    Male           NaN   
9455                22.0            NaN    Male           NaN   

     Occupation_Professional_Category Type_of_Business_Industry_of_Employment  \
9446            Elementary Occupation                                     NaN   
9447                          Officer                                     NaN   
9448                          Officer                                     NaN   
9449                          Officer                                     NaN   
9450                Semi-Professional                                     NaN   
9451         Service and Sales Worker                                     NaN   
9452         Service and Sales Worker                                     NaN   
9453            Elementary Occupation                                     NaN   
9454            Elementary Occupation                                     NaN   
9455                         Employee                                     NaN   

     Employment_Status  Time_at_Current_Employment_(Months)  \
9446          Contract                                  NaN   
9447         Full-Time                                115.0   
9448         Full-Time                                 41.0   
9449         Full-Time                                134.0   
9450         Full-Time                                 13.0   
9451         Full-Time                                 48.0   
9452         Full-Time                                 90.0   
9453          Contract                                146.0   
9454          Contract                                 12.0   
9455         Full-Time                                 31.0   

      Time_Employed_with_Previous_Employer_(Months)  Number_of_Dependents  \
9446                                            NaN                   NaN   
9447                                            NaN                   NaN   
9448                                            NaN                   NaN   
9449                                            NaN                   NaN   
9450                                            NaN                   NaN   
9451                                            NaN                   NaN   
9452                                            NaN                   NaN   
9453                                            NaN                   NaN   
9454                                            NaN                   NaN   
9455                                            NaN                   NaN   

     Loan_Purpose  Monthly_Income Government_Employee  
9446        Other         1463.56               False  
9447        Other         3657.43                

In [802]:


# Create function to fix dates
def fix_date_format(date_str):
    # Check 4th character, if '-' remove it, otherwise return original string
    return re.sub(r'^(...)-', r'\1', date_str)

# Apply the function to fix the date column
app_data['Application_Date'] = app_data['Application_Date'].apply(fix_date_format)

app_data[9446:9456]

Application_ID Application_Date   Loan_Approval_Status  \
9446     C302109447       2021-03-06                    NaN   
9447     C302109448       2018-07-05                Decline   
9448     C302109449       2021-02-21  Customer Cancellation   
9449     C302109450       2018-08-22  Customer Cancellation   
9450     C302109451       2018-09-19  Customer Cancellation   
9451     C302109452       2021-02-27                    NaN   
9452     C302109453       2018-06-17  Customer Cancellation   
9453     C302109454       2018-11-25                Decline   
9454     C302109455       2017-04-22  Customer Cancellation   
9455     C302109456       2021-08-12                    NaN   

      Age_at_Application Marital_Status  Gender Own_Rent_Home  \
9446                22.0            NaN  Female           NaN   
9447                31.0            NaN    Male           NaN   
9448                24.0            NaN    Male           NaN   
9449                37.0            NaN  Female           NaN   
9450                27.0            NaN    Male           NaN   
9451                39.0            NaN  Female           NaN   
9452                29.0            NaN    Male           NaN   
9453                37.0            NaN    Male           NaN   
9454                22.0            NaN    Male           NaN   
9455                22.0            NaN    Male           NaN   

     Occupation_Professional_Category Type_of_Business_Industry_of_Employment  \
9446            Elementary Occupation                                     NaN   
9447                          Officer                                     NaN   
9448                          Officer                                     NaN   
9449                          Officer                                     NaN   
9450                Semi-Professional                                     NaN   
9451         Service and Sales Worker                                     NaN   
9452         Service and Sales Worker                                     NaN   
9453            Elementary Occupation                                     NaN   
9454            Elementary Occupation                                     NaN   
9455                         Employee                                     NaN   

     Employment_Status  Time_at_Current_Employment_(Months)  \
9446          Contract                                  NaN   
9447         Full-Time                                115.0   
9448         Full-Time                                 41.0   
9449         Full-Time                                134.0   
9450         Full-Time                                 13.0   
9451         Full-Time                                 48.0   
9452         Full-Time                                 90.0   
9453          Contract                                146.0   
9454          Contract                                 12.0   
9455         Full-Time                                 31.0   

      Time_Employed_with_Previous_Employer_(Months)  Number_of_Dependents  \
9446                                            NaN                   NaN   
9447                                            NaN                   NaN   
9448                                            NaN                   NaN   
9449                                            NaN                   NaN   
9450                                            NaN                   NaN   
9451                                            NaN                   NaN   
9452                                            NaN                   NaN   
9453                                            NaN                   NaN   
9454                                            NaN                   NaN   
9455                                            NaN                   NaN   

     Loan_Purpose  Monthly_Income Government_Employee  
9446        Other         1463.56               False  
9447        Other         3657.43                

In [803]:

def change_to_date(df, date_cols):
    
    # Loop through list to change column dtype to datetime
    for col in date_cols:
        df[col] = pd.to_datetime(df[col])


# list of date_cols 
app_data_date_cols = ['Application_Date']

#Call function
change_to_date(app_data, app_data_date_cols)

# Confirm     
app_data['Application_Date'].dtype
app_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9492 entries, 0 to 9491
Data columns (total 16 columns):
 #   Column                                         Non-Null Count  Dtype         
---  ------                                         --------------  -----         
 0   Application_ID                                 9492 non-null   object        
 1   Application_Date                               9492 non-null   datetime64[ns]
 2   Loan_Approval_Status                           9354 non-null   object        
 3   Age_at_Application                             9471 non-null   float64       
 4   Marital_Status                                 7712 non-null   object        
 5   Gender                                         9488 non-null   object        
 6   Own_Rent_Home                                  8188 non-null   object        
 7   Occupation_Professional_Category               9361 non-null   object        
 8   Type_of_Business_Industry_of_Employment        7210 non-nu

In [804]:
contract_data.info()
contract_data_date_cols = ["Contract_Start_Date", "Expected_Contract_End_Date", "Actual_Contract_End_Date"]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9374 entries, 0 to 9373
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Contract_ID                 9374 non-null   object 
 1   Application_ID              7598 non-null   object 
 2   Key                         9374 non-null   int64  
 3   Loan_Type                   9374 non-null   object 
 4   Contract_Start_Date         9374 non-null   object 
 5   Expected_Contract_End_Date  9374 non-null   object 
 6   Actual_Contract_End_Date    2593 non-null   object 
 7   Loan_Term                   9374 non-null   object 
 8   Total_Loan_Amount           9374 non-null   float64
dtypes: float64(1), int64(1), object(7)
memory usage: 659.2+ KB


In [805]:
# contract_data[contract_data['Contract_Start_Date']== "202-1-08-31"]
contract_data[contract_data['Actual_Contract_End_Date']== "202-1-08-31"]
# contract_data[5890:5892]

Contract_ID Application_ID  Key                Loan_Type  \
5891  CONT105892     C302101952    1  Unsecured Personal Loan   
8608  CONT108609            NaN    3  Unsecured Personal Loan   
8609  CONT108610            NaN    2  Unsecured Personal Loan   

     Contract_Start_Date Expected_Contract_End_Date Actual_Contract_End_Date  \
5891          2017-10-02                 2020-10-02              202-1-08-31   
8608          2018-12-18                 2020-12-31              202-1-08-31   
8609         202-1-03-19                 2020-03-31              202-1-08-31   

          Loan_Term  Total_Loan_Amount  
5891  36 - Month(s)            8440.80  
8608  24 - Month(s)            2813.60  
8609  12 - Month(s)            1172.33

In [ ]:
# list of date_cols 
contract_data_date_cols = ["Contract_Start_Date", "Expected_Contract_End_Date", "Actual_Contract_End_Date"]

contract_data["Contract_Start_Date"] = contract_data["Contract_Start_Date"].apply(fix_date_format)
contract_data["Expected_Contract_End_Date"] = contract_data["Expected_Contract_End_Date"].apply(fix_date_format)
# contract_data["Actual_Contract_End_Date"] = contract_data["Actual_Contract_End_Date"].apply(fix_date_format)

# Apply the function to fix the date column
# for col in contract_data_date_cols:
#     contract_data[col] = contract_data[col].apply(fix_date_format)

contract_data[5891:5892]
# #Call function
# change_to_date(contract_data, contract_data_date_cols)

# # Confirm     
# # app_data['Application_Date'].dtype
# contract_data.info()

Contract_ID Application_ID  Key                Loan_Type  \
5891  CONT105892     C302101952    1  Unsecured Personal Loan   

     Contract_Start_Date Expected_Contract_End_Date Actual_Contract_End_Date  \
5891          2017-10-02                 2020-10-02              202-1-08-31   

          Loan_Term  Total_Loan_Amount  
5891  36 - Month(s)             8440.8

In [811]:
ContractsSnapshot_data.isnull().sum()#['Snapshot Date'].isnull().sum()

Contract_ID                     0
Snapshot Date                   0
Outstanding Amount              0
Due Amount                      0
Number of Days Past Due    195213
Contract Status                 0
latest_snapshot                 0
dtype: int64

In [ ]:
# # Get max snapshot date per contract
# latest_snapshot = ContractsSnapshot_data.groupby("Contract_ID")["Snapshot Date"].transform('max')
# latest_snapshot

# # Fill missing Actual_End_Date with the latest snapshot date per Contract_ID
# df['Actual_End_Date'] = df['Actual_End_Date'].fillna(latest_snapshot)

0          2018-12-31
1          2007-09-30
2          2007-09-30
3          2007-09-30
4          2007-09-30
             ...     
274366    202-1-08-31
274367    202-1-08-31
274368    202-1-08-31
274369    202-1-08-31
274370    202-1-08-31
Name: Snapshot Date, Length: 274371, dtype: object

In [ ]:
# Merge data
merged_contract_data = contract_data.merge(ContractsSnapshot_data, on="Contract_ID")
# merged_contract_data.head(2)

# Get max snapshot date per contract
latest_snapshot = ContractsSnapshot_data.groupby("Contract_ID")["Snapshot Date"].transform('max')
latest_snapshot

# Fill missing Actual_End_Date with the latest snapshot date per Contract_ID
merged_contract_data['Actual_Contract_End_Date'].fillna(latest_snapshot, inplace=True)

# Confirm
merged_contract_data["Actual_Contract_End_Date"].isnull().sum()

0

In [820]:
merged_contract_data["Actual_Contract_End_Date"] = merged_contract_data["Actual_Contract_End_Date"].apply(fix_date_format)

In [821]:
merged_contract_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 152869 entries, 0 to 152868
Data columns (total 15 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   Contract_ID                 152869 non-null  object 
 1   Application_ID              121952 non-null  object 
 2   Key                         152869 non-null  int64  
 3   Loan_Type                   152869 non-null  object 
 4   Contract_Start_Date         152869 non-null  object 
 5   Expected_Contract_End_Date  152869 non-null  object 
 6   Actual_Contract_End_Date    152869 non-null  object 
 7   Loan_Term                   152869 non-null  object 
 8   Total_Loan_Amount           152869 non-null  float64
 9   Snapshot Date               152869 non-null  object 
 10  Outstanding Amount          152869 non-null  float64
 11  Due Amount                  152869 non-null  float64
 12  Number of Days Past Due     28246 non-null   float64
 13  Contract Statu

In [823]:
ContractsSnapshot_data[ContractsSnapshot_data['Number of Days Past Due'].isnull()]

Contract_ID Snapshot Date  Outstanding Amount  Due Amount  \
0       CONT109375    2018-12-31                0.00         0.0   
1       CONT109376    2006-08-31              215.08         0.0   
2       CONT109376    2006-09-30              215.08         0.0   
3       CONT109376    2006-10-31              215.08         0.0   
4       CONT109376    2006-11-30              215.08         0.0   
...            ...           ...                 ...         ...   
274366  CONT116311   202-1-08-31           117233.29         0.0   
274367  CONT116312   202-1-08-31            43789.21         0.0   
274368  CONT116313   202-1-08-31           160023.45         0.0   
274369  CONT116314   202-1-08-31           140679.95         0.0   
274370  CONT116315   202-1-08-31            70339.98         0.0   

        Number of Days Past Due Contract Status latest_snapshot  
0                           NaN          Closed     202-1-08-31  
1                           NaN          Active     202-1-08-31  
2                           NaN          Active     202-1-08-31  
3                           NaN          Active     202-1-08-31  
4                           NaN          Active     202-1-08-31  
...                         ...             ...             ...  
274366                      NaN          Active     202-1-08-31  
274367                      NaN          Active     202-1-08-31  
274368                      NaN          Active     202-1-08-31  
274369                      NaN          Active     202-1-08-31  
274370                      NaN          Active     202-1-08-31  

[195213 rows x 7 columns]

In [ ]:

# merged_contract_data[["Contract_ID", "Expected_Contract_End_Date", 
#                       "Actual_Contract_End_Date", "Snapshot Date" ]]
# merged_contract_data["Contract_ID"].duplicated().sum()
# pd.to_datetime(merged_contract_data['Actual_Contract_End_Date'])#.datetime()

Most values noticed that have null past due days have a balance of 0. so we fill in with $0$.

In [825]:
merged_contract_data['Number of Days Past Due'].fillna(0, inplace=True)
merged_contract_data.isnull().sum()

Contract_ID                       0
Application_ID                30917
Key                               0
Loan_Type                         0
Contract_Start_Date               0
Expected_Contract_End_Date        0
Actual_Contract_End_Date          0
Loan_Term                         0
Total_Loan_Amount                 0
Snapshot Date                     0
Outstanding Amount                0
Due Amount                        0
Number of Days Past Due           0
Contract Status                   0
latest_snapshot                   0
dtype: int64

## Selecting relevant columns/ dropping irrelevant columns
The relevant columns for loan default prediction include:

Relevant Columns:
* ApplicationData.csv: Loan_Approval_Status, Monthly_Income, Number_of_Dependents, Time_at_Current_Employment_(Months), Employment_Status
* ContractsData.csv: Loan_Type, Loan_Term, Total_Loan_Amount
* ContractsSnapshotData.csv: Outstanding Amount, Number of Days Past Due, Contract Status
* CRB Data.csv: CRB Score, CRB Grade
* Current and Savings Account Data.csv: Average Balance, Overdraft Limit, Sum of Monthly Credit Transactions, Number of Bounced Cheques


In [494]:
# Select Relevant columns
new_app_data= app_data[["Application_ID", "Loan_Approval_Status", "Monthly_Income", 
                        "Number_of_Dependents", "Time_at_Current_Employment_(Months)", 
                        "Employment_Status", "Government_Employee"]]

new_contract_data = contract_data[["Contract_ID", "Application_ID", 
                                   "Contract_Start_Date", 
                                   "Expected_Contract_End_Date", "Loan_Type", 
                                   "Loan_Term", "Total_Loan_Amount"]]

new_ContractsSnapshot_data = ContractsSnapshot_data[["Contract_ID", 
                                                     "Outstanding Amount", "Due Amount",
                                                     "Number of Days Past Due", 
                                                     "Contract Status"]]

new_crb_data = crb_data[["Application_ID", "CRB Score", "CRB Grade"]]

new_curr_savings_data = curr_savings_data[["Application ID", "Average Balance", 
                                           "Overdraft Limit", 
                                           "Number of Days in Debit during Month",
                                           "Number of Days in Credit during Month",
                                           "Sum of Monthly Debit Transactions",
                                           "Sum of Monthly Credit Transactions",
                                           "Sum of Bounced Cheque transactions", 
                                           "Number of Bounced Cheques",
                                           "Sum of Over the Counter Withdrawals",
                                           "Account Type"]]

## Check duplicates

In [495]:
# check duplicates on the datasets
# app_data.duplicated().sum()
new_app_data.duplicated().sum()

0

In [496]:
# contract_data.duplicated().sum()
new_contract_data.duplicated().sum()

0

In [497]:
ContractsSnapshot_data.duplicated().sum() # 84
new_ContractsSnapshot_data.duplicated().sum() # 15724 ??

15724

In [498]:
crb_data.duplicated().sum() # 0
new_crb_data.duplicated().sum() # 75??

75

In [499]:
# curr_savings_data.duplicated().sum() # 0
new_curr_savings_data.duplicated().sum() # 194,048??

194048

In [500]:
# curr_savings_data.info()

## Handling missing Values

In [501]:
app_data.isnull().sum()

Application_ID                                      0
Application_Date                                    0
Loan_Approval_Status                              138
Age_at_Application                                 21
Marital_Status                                   1780
Gender                                              4
Own_Rent_Home                                    1304
Occupation_Professional_Category                  131
Type_of_Business_Industry_of_Employment          2282
Employment_Status                                 135
Time_at_Current_Employment_(Months)               232
Time_Employed_with_Previous_Employer_(Months)    9194
Number_of_Dependents                             1814
Loan_Purpose                                       42
Monthly_Income                                      0
Government_Employee                               181
dtype: int64

### Drop columns with more than 50% missing values

In [502]:
def calc_missing_values_percentage(df):
    
    # Calculate % of missing values
    missing_percentage = df.isnull().sum() * 100 / len(df)
    
    # Filter only columns with  50% missing values
    missing_percentage = missing_percentage[missing_percentage > 50] 
    
    # Sort in descending order
    missing_percentage = missing_percentage.sort_values(ascending=False)  

    return pd.DataFrame({'Column': missing_percentage.index, 'Missing_Percentage': missing_percentage.values})


In [503]:
# Get missing values
missing_app_data_values = calc_missing_values_percentage(app_data)
print(missing_app_data_values)

                                          Column  Missing_Percentage
0  Time_Employed_with_Previous_Employer_(Months)           96.860514


In [504]:
# Drop the column
app_data.drop(columns='Time_Employed_with_Previous_Employer_(Months)', inplace=True)

In [505]:
app_data[app_data["Loan_Approval_Status"].isnull()]


Application_ID Application_Date Loan_Approval_Status  Age_at_Application  \
135      C302100136       2018-07-03                  NaN                43.0   
144      C302100145       2018-09-06                  NaN                47.0   
7472     C302107473      202-1-04-16                  NaN                34.0   
7482     C302107483       2017-11-23                  NaN                28.0   
7517     C302107518       2018-05-24                  NaN                32.0   
...             ...              ...                  ...                 ...   
9446     C302109447      202-1-03-06                  NaN                22.0   
9451     C302109452      202-1-02-27                  NaN                39.0   
9455     C302109456      202-1-08-12                  NaN                22.0   
9474     C302109475       2018-10-23                  NaN                37.0   
9477     C302109478       2018-04-29                  NaN                48.0   

     Marital_Status  Gender Own_Rent_Home Occupation_Professional_Category  \
135         Married  Female        Family                              NaN   
144             NaN    Male           NaN                              NaN   
7472      Unmarried    Male        Family                     Professional   
7482      Unmarried  Female        Family         Service and Sales Worker   
7517      Unmarried    Male          Rent                     Professional   
...             ...     ...           ...                              ...   
9446            NaN  Female           NaN            Elementary Occupation   
9451            NaN  Female           NaN         Service and Sales Worker   
9455            NaN    Male           NaN                         Employee   
9474            NaN  Female           NaN                     Professional   
9477            NaN    Male           NaN                              NaN   

     Type_of_Business_Industry_of_Employment Employment_Status  \
135                            Civil Service         Full-Time   
144                                      NaN         Full-Time   
7472                           Civil Service         Full-Time   
7482            Business Process Outsourcing         Full-Time   
7517          Transportation & Communication         Full-Time   
...                                      ...               ...   
9446                                     NaN          Contract   
9451                                     NaN         Full-Time   
9455                                     NaN         Full-Time   
9474                                     NaN               NaN   
9477                                     NaN               NaN   

      Time_at_Current_Employment_(Months)  Number_of_Dependents  \
135                                 108.0                   1.0   
144                                   NaN                   NaN   
7472                                165.0                   0.0   
7482                                 25.0                   1.0   
7517                                108.0                   1.0   
...                                   ...                   ...   
9446                                  NaN                   NaN   
9451                                 48.0                   NaN   
9455                                 31.0                   NaN   
9474                                  NaN                   NaN   
9477                                  NaN                   NaN   

            Loan_Purpose  Monthly_Income Government_Employee  
135   Debt Consolidation            0.02               False  
144   Debt Consolidation         1634.23               False  
7472  Debt Consolidation         9416.45                True  
7482               Other         1195.58               False  
7517  Debt Consolidation         4277.54               False  
...                  ...             ...                 ...  
9446               Other         1463.56               F

['2021-04-16', '2021-04-16', '2032-12-20']


In [161]:
missing_app_data_values = calc_missing_values_percentage(app_data)
print(missing_app_data_values)
# app_data.info()

                                     Column  Missing_Percentage
0   Type_of_Business_Industry_of_Employment           24.041298
1                      Number_of_Dependents           19.110830
2                            Marital_Status           18.752634
3                             Own_Rent_Home           13.737885
4       Time_at_Current_Employment_(Months)            2.444164
5                       Government_Employee            1.906869
6                      Loan_Approval_Status            1.453856
7                         Employment_Status            1.422250
8          Occupation_Professional_Category            1.380110
9                              Loan_Purpose            0.442478
10                       Age_at_Application            0.221239
11                                   Gender            0.042141


Type_of_Business_Industry_of_Employment           24.041298
1                      Number_of_Dependents           19.110830
2                            Marital_Status           18.752634
3                             Own_Rent_Home           13.737885
4       Time_at_Current_Employment_(Months)            2.444164
5                       Government_Employee            1.906869
6                      Loan_Approval_Status            1.453856
7                         Employment_Status            1.422250
8          Occupation_Professional_Categor

In [170]:
def get_numerical_columns(df):
    return df.select_dtypes(include=['number']).columns.tolist()

def get_categorical_columns(df):
    return df.select_dtypes(include=['object', 'category']).columns.tolist()


In [169]:
# get numerical columns
numerical_cols = get_numerical_columns(app_data)
print(numerical_cols)

['Age_at_Application', 'Time_at_Current_Employment_(Months)', 'Number_of_Dependents', 'Monthly_Income']


In [171]:
# Example usage
categorical_cols = get_categorical_columns(app_data)
print(categorical_cols)

['Application_ID', 'Application_Date', 'Loan_Approval_Status', 'Marital_Status', 'Gender', 'Own_Rent_Home', 'Occupation_Professional_Category', 'Type_of_Business_Industry_of_Employment', 'Employment_Status', 'Loan_Purpose', 'Government_Employee']


In [ ]:
app_data['Type_of_Business_Industry_of_Employment'].fillna("Unknown", inplace= True)
app_data['Number_of_Dependents'].fillna()


In [164]:
app_data['Number_of_Dependents'].value_counts()

0.0     3547
1.0     2245
2.0     1233
3.0      463
4.0      133
5.0       40
6.0       11
9.0        2
7.0        2
13.0       1
8.0        1
Name: Number_of_Dependents, dtype: int64

In [166]:
# app_data.info()

## Feature transformation - log transformation/OHE/etc.

## Feature selection - (PCA/LASSO)

# Exploration Data Analysis

Univariate Analysis

Histograms & KDE Plots: Distribution of numerical features
Box Plots: Detect outliers in features like Monthly Income, Loan Amount
Bar Charts: Visualize frequency distribution for categorical variables like Loan Type, Contract Status, CRB Grade

Bivariate & Multivariate Analysis
* Correlation Matrix: Find relationships between features
* Use df.corr() & visualize with heatmaps (seaborn.heatmap())
* Scatter Plots: Identify linear/non-linear relationships
Pair Plots: Detect hidden patterns between numeric features
GroupBy Analysis: Compare default rates across different CRB grades
